spring-ai-client-chat/src/main/java/org/springframework/ai/chat/client/advisor/PromptChatMemoryAdvisor.java

```diff
@@ -91,56 +92,67 @@ public static Builder builder(ChatMemory chatMemory) {
-				streamAdvisorChain, this::before);
-
-		return new MessageAggregator().aggregateChatClientResponse(chatClientResponses, this::after);
+	public Scheduler getScheduler() {
+		return this.scheduler;
 	}
 
-	private ChatClientRequest before(ChatClientRequest chatClientRequest) {
-		String conversationId = this.doGetConversationId(chatClientRequest.context());
-		int chatMemoryRetrieveSize = this.doGetChatMemoryRetrieveSize(chatClientRequest.context());
-
+	@Override
+	public ChatClientRequest before(ChatClientRequest chatClientRequest, AdvisorChain advisorChain) {
+		String conversationId = doGetConversationId(chatClientRequest.context());
 		// 1. Retrieve the chat memory for the current conversation.
-		List<Message> memoryMessages = this.getChatMemoryStore().get(conversationId, chatMemoryRetrieveSize);
+		List<Message> memoryMessages = this.chatMemory.get(conversationId);
+		logger.debug("[PromptChatMemoryAdvisor.before] Memory before processing for conversationId={}: {}",
+				conversationId, memoryMessages);
 
-		// 2. Processed memory messages as a string.
+		// 2. Process memory messages as a string.
 		String memory = memoryMessages.stream()
 			.filter(m -> m.getMessageType() == MessageType.USER || m.getMessageType() == MessageType.ASSISTANT)
 			.map(m -> m.getMessageType() + ":" + m.getText())
 			.collect(Collectors.joining(System.lineSeparator()));
 
-		// 2. Augment the system message.
+		// 3. Augment the system message.
 		SystemMessage systemMessage = chatClientRequest.prompt().getSystemMessage();
 		String augmentedSystemText = this.systemPromptTemplate
 			.render(Map.of("instructions", systemMessage.getText(), "memory", memory));
 
-		// 3. Create a new request with the augmented system message.
+		// 4. Create a new request with the augmented system message.
 		ChatClientRequest processedChatClientRequest = chatClientRequest.mutate()
 			.prompt(chatClientRequest.prompt().augmentSystemMessage(augmentedSystemText))
 			.build();
 
-		// 4. Add the new user message to the conversation memory.
-		UserMessage userMessage = processedChatClientRequest.prompt().getUserMessage();
-		this.getChatMemoryStore().add(conversationId, userMessage);
+		// 5. Add all user messages from the current prompt to memory (after system
+		// message is generated)
+		List<UserMessage> userMessages = chatClientRequest.prompt().getUserMessages();
+		for (UserMessage userMessage : userMessages) {
+			this.chatMemory.add(conversationId, userMessage);
+			logger.debug("[PromptChatMemoryAdvisor.before] Added USER message to memory for conversationId={}: {}",
+					conversationId, userMessage.getText());
+		}
 
 		return processedChatClientRequest;
 	}
 
-	private void after(ChatClientResponse chatClientResponse) {
+	/**
+	 * Get the conversation id for the current context.
+	 * @param context the context
+	 * @return the conversation id
+	 */
+	protected String doGetConversationId(Map<String, Object> context) {
```

Comment:
This can be removed in favour of `getConversationId()` provided by `BaseChatMemoryAdvisor`.

Reply:
done